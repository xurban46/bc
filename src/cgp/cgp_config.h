/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

// hard configuration - affects compilation (array sizes etc.)
#define CGP_COLS 8
#define CGP_ROWS 4
#define CGP_INPUTS 9
#define CGP_OUTPUTS 1
#define CGP_LBACK 1
