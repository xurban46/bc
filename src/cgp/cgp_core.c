/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include "cgp_core.h"
#include "../random.h"


typedef struct {
    int size;
    int *values;
} int_array;


static int_array _allowed_gene_vals[CGP_COLS];
static int _mutation_rate;
static ga_fitness_func_t _fitness_func;

#ifdef CGP_LIMIT_FUNCS
    static int _allowed_functions_list[] = {
        c255,
        identity,
        add,
        add_sat,
        inversion,
        max,
        min
    };

    static int_array _allowed_functions = {
        .size = 7,
        .values = _allowed_functions_list
    };
#endif


#ifdef TEST_RANDOMIZE
    #define TEST_RANDOMIZE_PRINTF(...) printf(__VA_ARGS__)
#else
    #define TEST_RANDOMIZE_PRINTF(...) /* printf(__VA_ARGS__) */
#endif



/**
 * Initialize CGP internals
 */
void cgp_init(int mutation_rate, ga_fitness_func_t fitness_func)
{
    _mutation_rate = mutation_rate;
    _fitness_func = fitness_func;

    // calculate allowed values of node inputs in each column
    for (int x = 0; x < CGP_COLS; x++) {
        // range of outputs which can be connected to node in i-th column
        // maximum is actually by 1 larger than maximum allowed value
        // (so there is `val < maximum` in the for loop below instead of `<=`)
        int minimum = CGP_ROWS * (x - CGP_LBACK) + CGP_INPUTS;
        if (minimum < CGP_INPUTS) minimum = CGP_INPUTS;
        int maximum = CGP_ROWS * x + CGP_INPUTS;

        int size = CGP_INPUTS + maximum - minimum;
        _allowed_gene_vals[x].size = size;
        _allowed_gene_vals[x].values = (int*) malloc(sizeof(int) * size);

        int key = 0;
        // primary inputs
        for (int val = 0; val < CGP_INPUTS; val++, key++) {
            _allowed_gene_vals[x].values[key] = val;
        }

        // nodes to the left
        for (int val = minimum; val < maximum; val++, key++) {
            _allowed_gene_vals[x].values[key] = val;
        }
    }


#ifdef TEST_INIT
    for (int x = 0; x < CGP_COLS; x++) {
        printf ("x = %d: ", x);
        for (int y = 0; y < _allowed_gene_vals[x].size; y++) {
            if (y > 0) printf(", ");
            printf ("%d", _allowed_gene_vals[x].values[y]);
        }
        printf("\n");
    }
#endif
}


/**
 * Deinitialize CGP internals
 */
void cgp_deinit()
{
    for (int x = 0; x < CGP_COLS; x++) {
        free(_allowed_gene_vals[x].values);
    }
}


/**
 * Create a new cgp population with given size
 * @param  mutation rate (in number of genes)
 * @param  population size
 * @return
 */
ga_pop_t cgp_init_pop(int pop_size)
{
    /* prepare methods vector */
    ga_func_vect_t methods = {
        .alloc_genome = cgp_alloc_genome,
        .free_genome = cgp_free_genome,
        .init_genome = cgp_randomize_genome,

        .fitness = _fitness_func,
        .offspring = cgp_offspring,
    };

    /* initialize GA */
    ga_pop_t pop = ga_create_pop(pop_size, CGP_PROBLEM_TYPE, methods);
    return pop;
}


/* chromosome *****************************************************************/


/**
 * Allocates memory for new CGP genome
 * @param chromosome
 */
void* cgp_alloc_genome()
{
    return malloc(sizeof(struct cgp_genome));
}


/**
 * Initializes CGP genome to random values
 * @param chromosome
 */
int cgp_randomize_genome(ga_chr_t chromosome)
{
    cgp_genome_t genome = (cgp_genome_t) chromosome->genome;

    for (int i = 0; i < CGP_CHR_LENGTH; i++) {
        cgp_randomize_gene(genome, i);
    }

    cgp_find_active_blocks(chromosome);
    chromosome->has_fitness = false;

    return 0;
}


/**
 * Deinitialize CGP genome
 * @param  chromosome
 * @return
 */
void cgp_free_genome(void *genome)
{
    free(genome);
}


/* mutation *******************************************************************/


/**
 * Replace gene on given locus with random alele
 * @param chr
 * @param gene
 * @return whether active node was changed or not (phenotype has changed)
 */
bool cgp_randomize_gene(cgp_genome_t genome, int gene)
{
    if (gene >= CGP_CHR_LENGTH)
        return false;

    if (gene < CGP_CHR_OUTPUTS_INDEX) {
        // mutating node input or function
        int node_index = gene / 3;
        int gene_index = gene % 3;
        int col = cgp_node_col(node_index);

        TEST_RANDOMIZE_PRINTF("gene %u, node %u, gene %u, col %u\n", gene, node_index, gene_index, col);

        if (gene_index == CGP_FUNC_INPUTS) {
            // mutating function
            #ifdef CGP_LIMIT_FUNCS
                genome->nodes[node_index].function = (cgp_func_t) rand_schoice(_allowed_functions.size, _allowed_functions.values);
            #else
                genome->nodes[node_index].function = (cgp_func_t) rand_range(0, CGP_FUNC_COUNT - 1);
            #endif
            TEST_RANDOMIZE_PRINTF("func 0 - %u\n", CGP_FUNC_COUNT - 1);
            return genome->nodes[node_index].is_active;

        } else {
            // mutating input
            genome->nodes[node_index].inputs[gene_index] = rand_schoice(_allowed_gene_vals[col].size, _allowed_gene_vals[col].values);
            TEST_RANDOMIZE_PRINTF("input choice from %u\n", _allowed_gene_vals[col].size);
            return genome->nodes[node_index].is_active;
        }

    } else {
        // mutating primary output connection
        int index = gene - CGP_CHR_OUTPUTS_INDEX;
        genome->outputs[index] = rand_range(CGP_INPUTS, CGP_INPUTS + CGP_NODES - 1);
        TEST_RANDOMIZE_PRINTF("out %u - %u\n", CGP_INPUTS, CGP_INPUTS + CGP_NODES - 1);
        return true;
    }
}


/**
 * Mutate given chromosome
 * @param chr
 */
void cgp_mutate_chr(ga_chr_t chromosome)
{
    assert(_mutation_rate <= CGP_CHR_LENGTH);
    cgp_genome_t genome = (cgp_genome_t) chromosome->genome;

    int genes_to_change = rand_range(0, _mutation_rate);
    for (int i = 0; i < genes_to_change; i++) {
        int gene = rand_range(0, CGP_CHR_LENGTH - 1);
        cgp_randomize_gene(genome, gene);
    }

    cgp_find_active_blocks(chromosome);
    chromosome->has_fitness = false;
}


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_copy_genome(void *_dst, void *_src)
{
    cgp_genome_t dst = (cgp_genome_t) _dst;
    cgp_genome_t src = (cgp_genome_t) _src;

    memcpy(dst->nodes, src->nodes, sizeof(cgp_node_t) * CGP_NODES);
    memcpy(dst->outputs, src->outputs, sizeof(int) * CGP_OUTPUTS);
}


#define SWAP(A, B) (((A & 0x0F) << 4) | ((B & 0x0F)))
#define ADD_SAT(A, B) ((A > 0xFF - B) ? 0xFF : A + B)
#define MAX(A, B) ((A > B) ? A : B)
#define MIN(A, B) ((A < B) ? A : B)


/* evaluation *****************************************************************/


/**
 * Calculate output of given chromosome and inputs
 * @param chr
 */
void cgp_get_output(ga_chr_t chromosome, cgp_value_t *inputs, cgp_value_t *outputs)
{
    cgp_genome_t genome = (cgp_genome_t) chromosome->genome;
    cgp_value_t inner_outputs[CGP_INPUTS + CGP_NODES];

    // copy primary inputs to working array
    memcpy(inner_outputs, inputs, sizeof(cgp_value_t) * CGP_INPUTS);

    for (int i = 0; i < CGP_NODES; i++) {
        cgp_node_t *n = &(genome->nodes[i]);

        // skip inactive blocks
        if (!n->is_active) continue;

        cgp_value_t A = inner_outputs[n->inputs[0]];
        cgp_value_t B = inner_outputs[n->inputs[1]];
        cgp_value_t Y;

        switch (n->function) {
            case c255:          Y = 255;            break;
            case identity:      Y = A;              break;
            case inversion:     Y = 255 - A;        break;
            case b_or:          Y = A | B;          break;
            case b_not1or2:     Y = ~A | B;         break;
            case b_and:         Y = A & B;          break;
            case b_nand:        Y = ~(A & B);       break;
            case b_xor:         Y = A ^ B;          break;
            case rshift1:       Y = A >> 1;         break;
            case rshift2:       Y = A >> 2;         break;
            case swap:          Y = SWAP(A, B);     break;
            case add:           Y = A + B;          break;
            case add_sat:       Y = ADD_SAT(A, B);  break;
            case avg:           Y = (A + B) >> 1;   break;
            case max:           Y = MAX(A, B);      break;
            case min:           Y = MIN(A, B);      break;
            default:            abort();
        }

        inner_outputs[CGP_INPUTS + i] = Y;
    }

    for (int i = 0; i < CGP_OUTPUTS; i++) {
        outputs[i] = inner_outputs[genome->outputs[i]];
    }

#ifdef TEST_EVAL
    for (int i = 0; i < CGP_INPUTS + CGP_NODES; i++) {
        printf("%c: %u = %u\n", (i < CGP_INPUTS? 'I' : 'N'), i, inner_outputs[i]);
    }
    for (int i = 0; i < CGP_OUTPUTS; i++) {
        printf("O: %u = %u\n", i, outputs[i]);
    }
#endif
}


/**
 * Finds which blocks are active.
 * @param chromosome
 * @param active
 */
void cgp_find_active_blocks(ga_chr_t chromosome)
{
    cgp_genome_t genome = (cgp_genome_t) chromosome->genome;

    // first mark all nodes as inactive
    for (int i = CGP_NODES - 1; i >= 0; i--) {
        genome->nodes[i].is_active = false;
    }

    // mark inputs of primary outputs as active
    for (int i = 0; i < CGP_OUTPUTS; i++) {
        int index = genome->outputs[i] - CGP_INPUTS;
        // index may be negative (primary input), so do the check...
        if (index >= 0) {
            genome->nodes[index].is_active = true;
        }
    }

    // then walk nodes backwards and mark inputs of active nodes
    // as active nodes
    for (int i = CGP_NODES - 1; i >= 0; i--) {
        if (!genome->nodes[i].is_active) continue;
        cgp_node_t *n = &(genome->nodes[i]);

        for (int k = 0; k < CGP_FUNC_INPUTS; k++) {
            int index = n->inputs[k] - CGP_INPUTS;
            // index may be negative (primary input), so do the check...
            if (index >= 0) {
                genome->nodes[index].is_active = true;
            }
        }
    }
}



/* population *****************************************************************/


/**
 * Create new generation
 * @param pop
 * @param mutation_rate
 */
void cgp_offspring(ga_pop_t pop)
{
    ga_chr_t parent = pop->best_chromosome;

    #pragma omp parallel for
    for (int i = 0; i < pop->size; i++) {
        ga_chr_t chr = pop->chromosomes[i];
        if (chr == parent) continue;
        ga_copy_chr(chr, parent, cgp_copy_genome);
        cgp_mutate_chr(chr);
    }
}
