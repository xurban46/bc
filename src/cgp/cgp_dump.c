/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#include <stdlib.h>
#include <stdarg.h>

#include "cgp_dump.h"


#define FITNESS_PRECISE_FMT "%a"


static inline const char* cgp_func_name(cgp_func_t f) {
    const char *func_names[] = {
        " FF ",     // 255
        "  a ",     // a
        "FF-a",     // 255 - a
        " or ",     // a or b
        "~1|2",     // (not a) or b
        " and",     // a and b
        "nand",     // not (a and b)
        " xor",     // a xor b
        "a>>1",     // a >> 1
        "a>>2",     // a >> 2
        "swap",     // a <-> b
        " +  ",     // a + b
        " +S ",     // a +S b
        " avg",     // (a + b) >> 1
        " max",     // max(a, b)
        " min",     // min(a, b)
    };
    return func_names[f];
}


/**
 * Dumps chromosome to given file pointer
 * @param fp
 */
void cgp_dump_chr(ga_chr_t chr, FILE *fp, cgp_dump_format fmt)
{
    if (fmt == readable) {
        cgp_dump_chr_readable(chr, fp);

    } else if (fmt == asciiart) {
        cgp_dump_chr_asciiart(chr, fp, false);

    } else if (fmt == asciiart_active) {
        cgp_dump_chr_asciiart(chr, fp, true);

    } else {
        cgp_dump_chr_compat(chr, fp);
    }

}


/**
 * Dumps primary outputs to given file in "(X, Y, Z)"" format
 * @param chr
 * @param fp
 */
void cgp_dump_chr_outputs(ga_chr_t chr, FILE *fp)
{
    cgp_genome_t genome = (cgp_genome_t) chr->genome;

    fprintf(fp, "(");
    for (int i = 0; i < CGP_OUTPUTS; i++) {
        if (i > 0) fprintf(fp, ", ");
        fprintf(fp, "%u", genome->outputs[i]);
    }
    fprintf(fp, ")\n");
}


/**
 * Dumps basic info in human-friendly format
 * @param chr
 * @param fp
 */
void cgp_dump_chr_header(ga_chr_t chr, FILE *fp)
{
    fprintf(fp,
        "Inputs: %u\n"
        "Outputs: %u\n"
        "Size: %u x %u\n"
        "Blocks: %u-ary, %u output(s), %u functions\n",
        CGP_INPUTS, CGP_OUTPUTS, CGP_COLS, CGP_ROWS, CGP_FUNC_INPUTS,
        1, CGP_FUNC_COUNT);

    if (chr->has_fitness) {
        fprintf(fp, "Fitness: %lf\n", chr->fitness);
    } else {
        fprintf(fp, "Fitness: <none>\n");
    }
}


/**
 * Dumps chromosome to given file in CGP-viewer compatible format
 * @param chr
 * @param fp
 */
void cgp_dump_chr_compat(ga_chr_t chr, FILE *fp)
{
    cgp_genome_t genome = (cgp_genome_t) chr->genome;

    fprintf(fp, "{%u, %u, %u, %u, %u, %u, %u}",
        CGP_INPUTS, CGP_OUTPUTS, CGP_COLS, CGP_ROWS, CGP_FUNC_INPUTS,
        1, CGP_FUNC_COUNT);

    for (int i = 0; i < CGP_NODES; i++) {
        cgp_node_t *n = &(genome->nodes[i]);
        fprintf(fp, "([%u] %u, %u, %u)",
            CGP_INPUTS + i, n->inputs[0], n->inputs[1], n->function);
    }

    cgp_dump_chr_outputs(chr, fp);
}


/**
 * Dumps chromosome to given file in more human-friendly fashion
 * @param chr
 * @param fp
 */
void cgp_dump_chr_readable(ga_chr_t chr, FILE *fp)
{
    cgp_genome_t genome = (cgp_genome_t) chr->genome;

    cgp_dump_chr_header(chr, fp);

    for (int y = 0; y < CGP_ROWS; y++) {
        for (int x = 0; x < CGP_COLS; x++) {
            int i = cgp_node_index(x, y);
            cgp_node_t *n = &(genome->nodes[i]);

            fprintf(fp, "([%2u] %2u, %2u, %2u)  ",
                CGP_INPUTS + i, n->inputs[0], n->inputs[1], n->function);
        }
        if (CGP_OUTPUTS <= CGP_ROWS && y < CGP_OUTPUTS) {
            fprintf(fp, "  (%2u)", genome->outputs[y]);
        }
        fprintf(fp, "\n");
    }

    if (CGP_OUTPUTS > CGP_ROWS) {
        fprintf(fp, "Primary outputs: ");
        cgp_dump_chr_outputs(chr, fp);
    }
}


/**
 * ASCII-art rendering helper: renders primary input
 * @param in_counter
 * @param chr
 * @param fp
 */
void cgp_dump_chr_asciiart_input(int *in_counter, ga_chr_t chr, FILE *fp)
{
    if (*in_counter < CGP_INPUTS) fprintf(fp, "[%2u]>| ", (*in_counter)++);
    else fprintf(fp, "     | ");
}


/**
 * ASCII-art rendering helper: renders primary output
 * @param in_counter
 * @param chr
 * @param fp
 */
void cgp_dump_chr_asciiart_output(int *out_counter, ga_chr_t chr, FILE *fp)
{
    cgp_genome_t genome = (cgp_genome_t) chr->genome;
    if (*out_counter < CGP_OUTPUTS) fprintf(fp, ">[%2u]", genome->outputs[(*out_counter)++]);
}


/**
 * Dumps chromosome to given file as an ASCII-art, which looks like this:
 *
 *      .------------------------------------.
 *      |      .----.            .----.      |
 * [ 0]>| [ 1]>|    |>[ 4]  [ 0]>|    |>[ 6] |>[ 6]
 * [ 1]>| [ 3]>|  a |       [ 4]>| xor|      |
 * [ 2]>|      '----'            '----'      |
 * [ 3]>|      .----.            .----.      |
 *      | [ 3]>|    |>[ 5]  [ 4]>|    |>[ 7] |
 *      | [ 0]>|a>>1|       [ 5]>|nand|      |
 *      |      '----'            '----'      |
 *      '------------------------------------'
 *
 * @param chr
 * @param fp
*/
void cgp_dump_chr_asciiart(ga_chr_t chr, FILE *fp, bool only_active_blocks)
{
    int in_counter = 0;
    int out_counter = 0;
    cgp_genome_t genome = (cgp_genome_t) chr->genome;

    cgp_dump_chr_header(chr, fp);

    /* top of the circuit */
    fprintf(fp, "     .--");
    for (int x = 0; x < CGP_COLS; x++) {
        fprintf(fp, "----------------");
        if (x == CGP_COLS - 1) fprintf(fp, ".\n");
        else fprintf(fp, "--");
    }

    for (int y = 0; y < CGP_ROWS; y++) {
        if (y != 0) cgp_dump_chr_asciiart_input(&in_counter, chr, fp);
        else fprintf(fp, "     | ");

        /* top of the blocks */
        fprintf(fp, " ");
        for (int x = 0; x < CGP_COLS; x++) {
            int i = cgp_node_index(x, y);
            cgp_node_t *n = &(genome->nodes[i]);

            if (only_active_blocks && !n->is_active) {
                fprintf(fp, "                ");
            } else {
                fprintf(fp, "    .----.      ");
            }

            if (x == CGP_COLS - 1) fprintf(fp, "|");
            else fprintf(fp, "  ");
        }

        if (y != 0) cgp_dump_chr_asciiart_output(&out_counter, chr, fp);
        fprintf(fp, "\n");
        cgp_dump_chr_asciiart_input(&in_counter, chr, fp);

        /* first line of blocks */
        for (int x = 0; x < CGP_COLS; x++) {
            int i = cgp_node_index(x, y);
            cgp_node_t *n = &(genome->nodes[i]);

            if (only_active_blocks && !n->is_active) {
                fprintf(fp, "                ");
            } else {
                fprintf(fp, "[%2u]>|    |>[%2u]", n->inputs[0], CGP_INPUTS + i);
            }

            if (x == CGP_COLS - 1) fprintf(fp, " |");
            else fprintf(fp, "  ");
        }

        cgp_dump_chr_asciiart_output(&out_counter, chr, fp);
        fprintf(fp, "\n");
        cgp_dump_chr_asciiart_input(&in_counter, chr, fp);

        /* second line of blocks */
        for (int x = 0; x < CGP_COLS; x++) {
            int i = cgp_node_index(x, y);
            cgp_node_t *n = &(genome->nodes[i]);

            if (only_active_blocks && !n->is_active) {
                fprintf(fp, "                ");
            } else {
                fprintf(fp, "[%2u]>|%s|     ", n->inputs[1], cgp_func_name(n->function));
            }

            if (x == CGP_COLS - 1) fprintf(fp, " |");
            else fprintf(fp, "  ");
        }

        cgp_dump_chr_asciiart_output(&out_counter, chr, fp);
        fprintf(fp, "\n");
        cgp_dump_chr_asciiart_input(&in_counter, chr, fp);

        /* bottom of the blocks */
        fprintf(fp, " ");
        for (int x = 0; x < CGP_COLS; x++) {
            int i = cgp_node_index(x, y);
            cgp_node_t *n = &(genome->nodes[i]);

            if (only_active_blocks && !n->is_active) {
                fprintf(fp, "                ");
            } else {
                fprintf(fp, "    '----'      ");
            }

            if (x == CGP_COLS - 1) fprintf(fp, "|");
            else fprintf(fp, "  ");
        }

        cgp_dump_chr_asciiart_output(&out_counter, chr, fp);
        fprintf(fp, "\n");
    }

    /* bottom of the circuit */
    fprintf(fp, "     '--");
    for (int x = 0; x < CGP_COLS; x++) {
        fprintf(fp, "----------------");
        if (x == CGP_COLS - 1) fprintf(fp, "'\n");
        else fprintf(fp, "--");
    }
}


/**
 * Dumps whole population to given file pointer with chromosomes in
 * CGP-viewer compatible format
 * @param pop
 * @param fp
 */
void cgp_dump_pop_compat(ga_pop_t pop, FILE *fp)
{
    fprintf(fp, "Generation: %d\n", pop->generation);
    fprintf(fp, "Best chromosome: %d\n", pop->best_chr_index);
    fprintf(fp, "Chromosomes: %d\n", pop->size);

    for (int i = 0; i < pop->size; i++) {
        ga_chr_t chr = pop->chromosomes[i];
        if (chr->has_fitness) {
            fprintf(fp, "Fitness: Y " FITNESS_PRECISE_FMT "\n", chr->fitness);
        } else {
            fprintf(fp, "Fitness: N\n");
        }
        cgp_dump_chr_compat(chr, fp);
    }
}
