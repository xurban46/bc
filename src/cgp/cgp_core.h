/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include "../ga.h"
#include "cgp_config.h"


#define CGP_FUNC_INPUTS 2
#define CGP_NODES (CGP_COLS * CGP_ROWS)
#define CGP_CHR_OUTPUTS_INDEX ((CGP_FUNC_INPUTS + 1) * CGP_NODES)
#define CGP_CHR_LENGTH (CGP_CHR_OUTPUTS_INDEX + CGP_OUTPUTS)

static const ga_problem_type_t CGP_PROBLEM_TYPE = maximize;

typedef unsigned char cgp_value_t;

#define CGP_FUNC_COUNT 16
typedef enum
{
    c255 = 0,   // 255
    identity,   // a
    inversion,  // 255 - a
    b_or,       // a or b
    b_not1or2,  // (not a) or b
    b_and,      // a and b
    b_nand,     // not (a and b)
    b_xor,      // a xor b
    rshift1,    // a >> 1
    rshift2,    // a >> 2
    swap,       // a <-> b
    add,        // a + b
    add_sat,    // a +S b
    avg,        // (a + b) >> 1
    max,        // max(a, b)
    min,        // min(a, b)
} cgp_func_t;


/**
 * One CGP node (function block)
 */
typedef struct {
    int inputs[CGP_FUNC_INPUTS];
    cgp_func_t function;
    bool is_active;
} cgp_node_t;


/**
 * Chromosome
 */
struct cgp_genome {
    cgp_node_t nodes[CGP_COLS * CGP_ROWS];
    int outputs[CGP_OUTPUTS];
};
typedef struct cgp_genome* cgp_genome_t;


/**
 * Initialize CGP internals
 */
void cgp_init(int mutation_rate, ga_fitness_func_t fitness_func);


/**
 * Deinitialize CGP internals
 */
void cgp_deinit();


/**
 * Create a new CGP population with given size
 * @param  mutation rate (in number of genes)
 * @return
 */
ga_pop_t cgp_init_pop(int pop_size);


/**
 * Allocates memory for new CGP genome
 * @return pointer to newly allocated genome
 */
void* cgp_alloc_genome();


/**
 * Initializes CGP genome to random values
 * @param chromosome
 * @return 0 on success, other value on error
 */
int cgp_randomize_genome(ga_chr_t chromosome);


/**
 * Deinitialize CGP genome
 * @param  genome
 * @return
 */
void cgp_free_genome(void *genome);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_copy_genome(void *_dst, void *_src);


/**
 * Replace gene on given locus with random alele
 * @param chr
 * @param gene
 * @return whether active node was changed or not (phenotype has changed)
 */
bool cgp_randomize_gene(cgp_genome_t genome, int gene);


/**
 * Mutate given chromosome
 * @param chr
 */
void cgp_mutate_chr(ga_chr_t chromosome);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_replace_chr(ga_chr_t chromosome, ga_chr_t replacement);


/**
 * Calculate output of given chromosome and inputs
 * @param chr
 */
void cgp_get_output(ga_chr_t chromosome, cgp_value_t *inputs, cgp_value_t *outputs);


/**
 * Create new generation
 * @param pop
 * @param mutation_rate
 */
void cgp_offspring(ga_pop_t pop);


/**
 * Returns index of node in given column and row
 * @param  col
 * @param  row
 * @return
 */
static inline int cgp_node_index(int col, int row)
{
    return CGP_ROWS * col + row;
}


/**
 * Returns node column based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_col(int index)
{
    return index / CGP_ROWS;
}


/**
 * Returns node row based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_row(int index)
{
    return index % CGP_ROWS;
}


/**
 * Finds which blocks are active.
 * @param chromosome
 * @param active
 */
void cgp_find_active_blocks(ga_chr_t chromosome);
