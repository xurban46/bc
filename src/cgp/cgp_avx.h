/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include <immintrin.h>

#include "cgp_core.h"


typedef __m256i __m256i_aligned __attribute__ ((aligned (32)));


/**
 * Calculate output of given chromosome and inputs using AVX instructions
 * @param chr
 * @param inputs
 * @param outputs
 */
void cgp_get_output_avx(ga_chr_t chromosome, __m256i_aligned inputs[CGP_INPUTS], __m256i_aligned outputs[CGP_OUTPUTS]);
