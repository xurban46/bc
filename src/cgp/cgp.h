/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include "cgp_core.h"
#include "cgp_dump.h"
#include "cgp_load.h"
