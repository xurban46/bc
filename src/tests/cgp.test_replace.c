/**
 * Tests CGP chromosome replacement.
 */


#include <stdlib.h>
#include <stdio.h>

#include "../cgp.h"


int main(int argc, char const *argv[])
{
    cgp_init();

    cgp_genome_t genome1 = (cgp_genome_t) malloc(sizeof(struct cgp_genome));
    struct ga_chr chr1 = {
        .genome = genome1
    };

    cgp_genome_t genome2 = (cgp_genome_t) malloc(sizeof(struct cgp_genome));
    struct ga_chr chr2 = {
        .genome = genome2
    };

    cgp_init_chr(&chr1);
    cgp_init_chr(&chr2);

    cgp_dump_chr_compat(&chr1, stdout);
    printf("\n");
    cgp_dump_chr_compat(&chr2, stdout);
    printf("\n");

    cgp_replace_chr(&chr2, &chr1);
    cgp_dump_chr_compat(&chr2, stdout);

    free(genome1);
    free(genome2);
    cgp_deinit();
}
