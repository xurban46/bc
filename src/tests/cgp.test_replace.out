{9, 1, 8, 4, 2, 1, 16}([9] 1, 7, 9)([10] 7, 5, 15)([11] 1, 3, 9)([12] 1, 5, 11)([13] 0, 4, 3)([14] 1, 9, 2)([15] 10, 11, 11)([16] 8, 3, 13)([17] 3, 5, 14)([18] 3, 14, 15)([19] 13, 16, 6)([20] 13, 5, 7)([21] 18, 5, 3)([22] 19, 1, 13)([23] 3, 1, 8)([24] 5, 5, 4)([25] 4, 2, 13)([26] 23, 0, 4)([27] 2, 3, 14)([28] 6, 6, 1)([29] 26, 2, 12)([30] 27, 1, 9)([31] 4, 27, 1)([32] 0, 28, 7)([33] 0, 2, 14)([34] 5, 29, 12)([35] 7, 1, 15)([36] 30, 32, 10)([37] 1, 3, 11)([38] 0, 33, 15)([39] 34, 7, 12)([40] 33, 7, 12)(10)

{9, 1, 8, 4, 2, 1, 16}([9] 0, 3, 3)([10] 0, 1, 10)([11] 1, 8, 9)([12] 5, 2, 5)([13] 5, 9, 4)([14] 3, 10, 14)([15] 6, 1, 9)([16] 8, 10, 8)([17] 8, 6, 5)([18] 4, 1, 1)([19] 13, 2, 15)([20] 0, 5, 9)([21] 5, 17, 5)([22] 3, 5, 11)([23] 20, 0, 7)([24] 0, 3, 15)([25] 4, 21, 8)([26] 1, 5, 4)([27] 1, 2, 3)([28] 7, 8, 14)([29] 7, 5, 3)([30] 1, 5, 4)([31] 3, 26, 14)([32] 2, 3, 10)([33] 0, 2, 2)([34] 7, 2, 15)([35] 32, 8, 13)([36] 4, 30, 8)([37] 35, 0, 12)([38] 36, 33, 2)([39] 2, 3, 11)([40] 34, 4, 11)(9)

{9, 1, 8, 4, 2, 1, 16}([9] 1, 7, 9)([10] 7, 5, 15)([11] 1, 3, 9)([12] 1, 5, 11)([13] 0, 4, 3)([14] 1, 9, 2)([15] 10, 11, 11)([16] 8, 3, 13)([17] 3, 5, 14)([18] 3, 14, 15)([19] 13, 16, 6)([20] 13, 5, 7)([21] 18, 5, 3)([22] 19, 1, 13)([23] 3, 1, 8)([24] 5, 5, 4)([25] 4, 2, 13)([26] 23, 0, 4)([27] 2, 3, 14)([28] 6, 6, 1)([29] 26, 2, 12)([30] 27, 1, 9)([31] 4, 27, 1)([32] 0, 28, 7)([33] 0, 2, 14)([34] 5, 29, 12)([35] 7, 1, 15)([36] 30, 32, 10)([37] 1, 3, 11)([38] 0, 33, 15)([39] 34, 7, 12)([40] 33, 7, 12)(10)
